src/config.py

```diff
@@ -0,0 +1,176 @@
+# value. For example
+#
+# import spras.config as config
+# container_framework = config.config.container_framework
+#
+# will grab the top level registry configuration option as it appears in the config file
+
+import copy as copy
+import itertools as it
+import os
+
+import numpy as np
+import yaml
+
+from src.util import hash_params_sha1_base32
+
+# The default length of the truncated hash used to identify parameter combinations
+DEFAULT_HASH_LENGTH = 7
+DEFAULT_CONTAINER_PREFIX = "docker.io/reedcompbio"
+
+config = None
+
+# This will get called in the Snakefile, instantiating the singleton with the raw config
+def init_global(config_dict):
+    global config
+    config = Config(config_dict)
+
+def init_from_file(filepath):
+    global config
+
+    # Handle opening the file and parsing the yaml
+    filepath = os.path.abspath(filepath)
+    try:
+        with open(filepath, 'r') as yaml_file:
+            config_dict = yaml.safe_load(yaml_file)
+    except FileNotFoundError:
+        print(f"Error: The specified config '{filepath}' could not be found.")
+        return False
+    except yaml.YAMLError as e:
+        print(f"Error: Failed to parse config '{filepath}': {e}")
+        return False
+
+    # And finally, initialize
+    config = Config(config_dict)
+
+
+class Config:
+    def __init__(self, raw_config):
+        # Since process_config winds up modifying the raw_config passed to it as a side effect,
+        # we'll make a deep copy here to guarantee we don't break anything. This preserves the
+        # config as it's given to the Snakefile by Snakemake
+        _raw_config = copy.deepcopy(raw_config)
+        self.process_config(_raw_config)
+
+    def process_config(self, raw_config):
+        if raw_config == {}:
+            raise ValueError("Config file cannot be empty. Use --configfile <filename> to set a config file.")
+
+        # Set up a few top-level config variables
+        self.out_dir = raw_config["reconstruction_settings"]["locations"]["reconstruction_dir"]
+
+        # We allow the container framework not to be defined in the config. In the case it isn't, default to docker.
+        # However, if we get a bad value, we raise an exception.
+        if "container_framework" in raw_config:
+            container_framework = raw_config["container_framework"].lower()
+            if container_framework not in ("docker", "singularity"):
+                msg = "SPRAS was configured to run with an unknown container framework: '" + raw_config["container_framework"] + "'. Accepted values are 'docker' or 'singularity"
+                raise ValueError(msg)
+            self.container_framework = container_framework
+        else:
+            self.container_framework = "docker"
+
+        # Grab registry from the config, and if none is provided default to docker
+        if "container_registry" in raw_config and raw_config["container_registry"]["base_url"] != "" and raw_config["container_registry"]["owner"] != "":
```

Comment:
Do we need any new config processing test cases to confirm logic like this works as expected? We didn't have good test coverage of the original config file parsing but could try to do better with new parts.